lib/basedriver/capabilities.js

```diff
@@ -0,0 +1,111 @@
+
+  // Remove the 'presence' constraint if we're not checking for it
+  if (skipPresenceConstraint) {
+    constraints = Object.assign(constraints);
+    for (let key of Object.keys(constraints)) {
+      delete constraints[key].presence;
+    }
+  }
+
+  let validationErrors = validator.validate(_.pickBy(caps, util.hasValue),
+                                              constraints,
+                                              {fullMessages: false});
+                                          
+  if (validationErrors) {
+    for (let [attribute, reasons] of _.toPairs(validationErrors)) {
+      for (let reason of reasons) {
+        message += ` ${attribute} ${reason},`;
+      }
+    }
+    message = `${message.slice(0, -1)}.`;
+    throw new Error(message);
+  }
+
+  // Return caps
+  return caps;
+}
+
+// Parse capabilities (based on https://www.w3.org/TR/webdriver/#processing-capabilities)
+function parseCaps (caps, constraints = {}, shouldValidateCaps = true) {
+  if (!_.isObject(caps)) {
+    throw new Error('The capabilities argument was not valid for the following reason(s): "capabilities" must be a JSON object.');
+  }
+
+  let {alwaysMatch: requiredCaps, firstMatch: allFirstMatchCaps = []} = caps; // jshint ignore:line
+
+  if (!_.isObject(requiredCaps)) {
+    requiredCaps = {}; 
+  }
+
+  requiredCaps = validateCaps(requiredCaps, constraints);
+
+  // Reject 'firstMatch' argument if it's not an array (see spec #3.2)
+  if (!_.isArray(allFirstMatchCaps)) {
+    throw new Error('The capabilities.firstMatch argument was not valid for the following reason(s): "capabilities.firstMatch" must be a JSON array or undefined');
+  }
+
+  // Validate the firstMatch capabilities one-by one (see spec #5)
+  let validatedFirstMatchCaps = allFirstMatchCaps.map((firstMatchCaps, index) => {
+    try {
+      return shouldValidateCaps ? validateCaps(firstMatchCaps, constraints) : firstMatchCaps;
+    } catch (e) {
+      throw new Error(`The capabilities.firstMatch argument was not valid for the following reason: "capabilities.firstMatch[${index}] ${e.message}`);
+    }
+  });
+
+  // Try to match requiredCaps with one of the validated firstMatch capabilities (see spec #6)
+  let matchedCaps = null;
+  for (let firstMatchCaps of validatedFirstMatchCaps) {
+    try {
+      let matchedCaps = mergeCaps(requiredCaps, firstMatchCaps);
+      if (matchedCaps) {
+        break;
+      }
+    } catch (ign) { }
+  }
+
+  // If there was no match, use the de-facto requiredCaps on its own
+  matchedCaps = matchedCaps || requiredCaps;
+  if (!matchedCaps) {
+    matchedCaps = requiredCaps;
+  }
+
+  // Returns variables for testing purposes
+  return {requiredCaps, allFirstMatchCaps, validatedFirstMatchCaps, matchedCaps};
+}
+
+// Calls parseCaps and just returns the matchedCaps variable
+function processCaps (caps, constraints = {}, shouldValidateCaps = true) {
+  return parseCaps(caps, constraints, shouldValidateCaps).matchedCaps;
```

Comment:
We only used matchedCaps. I probably could refactor it to only return that, the only reason I did so was because I was following the spec and wanted to test it all.